packages/flame/lib/src/components/shape_component.dart

```diff
@@ -1,35 +1,118 @@
 import 'dart:ui' hide Offset;
 
+import 'package:flutter/material.dart';
+
 import '../../components.dart';
 import '../../geometry.dart';
+import '../../palette.dart';
 import '../anchor.dart';
 import '../extensions/vector2.dart';
 
 class ShapeComponent extends PositionComponent {
-  final Shape shape;
-  final Paint shapePaint;
+  Shape? shape;
+  Paint paint = BasicPalette.white.paint();
 
-  ShapeComponent(
+  ShapeComponent({
     this.shape,
-    this.shapePaint, {
+    Paint? paint,
     Anchor anchor = Anchor.topLeft,
     int? priority,
   }) : super(
-          position: shape.position,
-          size: shape.size,
-          angle: shape.angle,
+          position: shape?.position,
+          size: shape?.size,
+          angle: shape?.angle ?? 0,
           anchor: anchor,
           priority: priority,
         ) {
-    shape.isCanvasPrepared = true;
+    this.paint = paint ?? this.paint;
+    shape?.isCanvasPrepared = true;
   }
 
   @override
   void render(Canvas canvas) {
     super.render(canvas);
-    shape.render(canvas, shapePaint);
+    shape?.render(canvas, paint);
   }
 
   @override
-  bool containsPoint(Vector2 point) => shape.containsPoint(point);
+  bool containsPoint(Vector2 point) => shape?.containsPoint(point) ?? false;
+
+  void updateShape(Shape shape) {
+    this.shape = shape..isCanvasPrepared = true;
+    position = shape.position;
+    size = shape.size;
+    angle = shape.angle;
+  }
+}
+
+class CircleComponent extends ShapeComponent {
```

Comment:
Separate publicly-visible classes should probably live in separate files.

Reply:
It is not uncommon to declare several classes in one file in Dart, especially when these classes barely do anything at all it would just pollute the tree of files.

https://dart.dev/guides/language/effective-dart/design#consider-declaring-multiple-classes-in-the-same-library (not exactly about this, but they kind of mention that it is not an anti-pattern.)